src/combinatorics/set_family/intersecting.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.fintype.basic
+import order.upper_lower
+
+/-!
+# Intersecting families
+
+This file defines intersecting families and proves their basic properties.
+
+## Main declarations
+
+* `set.intersecting`: Predicate for a set of elements in a generalized boolean algebra to be an
+  intersecting family.
+* `set.intersecting.card_le`: An intersecting family can only take up to half the elements, because
+  `a` and `aᶜ` cannot simultaneously be in it.
+* `set.intersecting.is_max_iff_card_eq`: Any maximal intersecting family takes up half the elements.
+
+## References
+
+* [D. J. Kleitman, *Families of non-disjoint subsets*][kleitman1966]
+-/
+
+open finset
+
+variables {α : Type*}
+
+namespace set
+section generalized_boolean_algebra
+variables [generalized_boolean_algebra α] {s t : set α} {a b c : α}
+
+/-- A set family is intersecting if every pair of elements is non-disjoint. -/
+def intersecting (s : set α) : Prop := ∀ ⦃a⦄, a ∈ s → ∀ ⦃b⦄, b ∈ s → ¬ disjoint a b
+
+@[mono] lemma intersecting.mono (h : t ⊆ s) (hs : s.intersecting) : t.intersecting :=
+λ a ha b hb, hs (h ha) (h hb)
+
+lemma intersecting.not_bot_mem (hs : s.intersecting) : ⊥ ∉ s := λ h, hs h h disjoint_bot_left
+
+lemma intersecting.ne_bot (hs : s.intersecting) (ha : a ∈ s) : a ≠ ⊥ :=
+ne_of_mem_of_not_mem ha hs.not_bot_mem
+
+lemma intersecting_empty : (∅ : set α).intersecting := λ _, false.elim
+
+@[simp] lemma intersecting_singleton : ({a} : set α).intersecting ↔ a ≠ ⊥ := by simp [intersecting]
+
+protected lemma subsingleton.intersecting (hs : s.subsingleton) : s.intersecting ↔ s ≠ {⊥} :=
+begin
+  refine ⟨_, λ h a ha b hb hab, h $ hs.eq_singleton_of_mem _⟩,
+  { rintro h rfl,
+    exact intersecting_singleton.1 h rfl },
+  { rw [hs ha hb, disjoint_self] at hab,
+    rwa ←hab }
+end
```

Comment:
On top of the previous, it might be nice to have 
```
lemma intersecting_iff_eq_empty_of_subsingleton [subsingleton α] (s : set α) :
  s.intersecting ↔ s = ∅ :=
```